### PR TITLE
APIs for reversed binary operators

### PR DESCRIPTION
**Pull request by pgovind**
To enable statements such as `1 - intColumn`

### CODE REVIEW

File: src/Microsoft.Data/DataFrame.BinaryOperations.cs [hunk 1]
```diff
@@ -26,12 +26,15 @@ public DataFrame Add<T>(IReadOnlyList<T> values, bool inPlace = false)
             for (int i = 0; i < Columns.Count; i++)
             {
                 DataFrameColumn baseColumn = _columnCollection[i];
+                baseColumn._reverseOrderOfBinaryOperations = _reverseOrderOfBinaryOperations;
```
Comment (pgovind) [inline src/Microsoft.Data/DataFrame.BinaryOperations.cs]: Adding internal state to `DataFrame` and `PrimitiveDataFrameColumn` seemed to be a cleaner than adding parameters to each binary op.

**Comment by pgovind** [comment]
Ok, can be reviewed now.

File: src/Microsoft.Data/DataFrame.BinaryOperations.cs (hunk 1 above)
Comment (eerhardt) [reply src/Microsoft.Data/DataFrame.BinaryOperations.cs]: I think I disagree. If an exception is thrown between here and line 35, this state is going to be left to this old value.

---
In reply to: [340205425](https://github.com/dotnet/corefxlab/pull/2769#discussion_r340205425) [](ancestors = 340205425)

Comment (eerhardt) [reply src/Microsoft.Data/DataFrame.BinaryOperations.cs]: Also, how could someone write their own derived column class? They won't be able to get access to this internal state.

Can we find some other way to enable this?

---
In reply to: [340322971](https://github.com/dotnet/corefxlab/pull/2769#discussion_r340322971) [](ancestors = 340322971,340205425)

Comment (pgovind) [reply src/Microsoft.Data/DataFrame.BinaryOperations.cs]: For the 2nd question, I thought of that and put the flag on the base class :)

I think I'm going to make new `Radd`, `Rdivide` etc virtual methods. That should fix all our concerns

Comment (eerhardt) [reply src/Microsoft.Data/DataFrame.BinaryOperations.cs]: The method names should follow the .NET guidelines though - 

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/naming-guidelines

File: src/Microsoft.Data/DataFrame.BinaryOperations.cs [hunk 2]
```diff
@@ -506,6 +504,123 @@ public DataFrame ElementwiseLessThan<T>(T value)
             }
             return retDataFrame;
         }
-        #endregion
+
+        public DataFrame RAdd<T>(T value, bool inPlace = false)
```
Comment (pgovind) [inline src/Microsoft.Data/DataFrame.BinaryOperations.cs]: All these methods need documentation

File: src/Microsoft.Data/DataFrameColumn.BinaryOperations.cs [hunk 3]
```diff
@@ -23,6 +76,12 @@ public virtual DataFrameColumn Add<T>(T value, bool inPlace = false)
             throw new NotImplementedException();
         }
 
+        public virtual DataFrameColumn RAdd<T>(T value, bool inPlace = false)
```
Comment (pgovind) [inline src/Microsoft.Data/DataFrameColumn.BinaryOperations.cs]: Needs documentation

File: src/Microsoft.Data/DataFrame.BinaryOperations.cs (hunk 2 above)
Comment (pgovind) [inline src/Microsoft.Data/DataFrame.BinaryOperations.cs]: `RAdd` follows our naming guidelines, but `ROr` sounds a little weird. I don't think we should name it `ReverseAdd` though

**Comment by pgovind** [comment]
Can be reviewed. This is much better than the previous version.

**Comment by pgovind** [comment]
https://github.com/dotnet/corefxlab/issues/2749 can be fixed only after this goes in.

File: src/Microsoft.Data/DataFrame.BinaryOperations.cs [hunk 4]
```diff
@@ -506,6 +552,147 @@ public DataFrame ElementwiseLessThan<T>(T value)
             }
             return retDataFrame;
         }
-        #endregion
+
+        /// <summary>
+        /// Performs a reversed element wise addition on each column
+        /// </summary>
+        public DataFrame RAdd<T>(T value, bool inPlace = false)
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrame.BinaryOperations.cs]: `RAdd` doesn't follow our naming guidelines. See [https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/general-naming-conventions](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/general-naming-conventions).

File: src/Microsoft.Data/DataFrame.BinaryOperators.cs [hunk 5]
```diff
@@ -16,339 +16,685 @@ public partial class DataFrame
...
+        }
 
         public static DataFrame operator &(DataFrame df, bool value)
         {
             return df.And(value);
         }
+
+        public static DataFrame operator &(bool value, DataFrame df)
+        {
+            DataFrame ret = df.RAnd(value, inPlace: false);
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrame.BinaryOperators.cs]: Why not just `return df.RAnd(value, inPlace: false);`?

File: src/Microsoft.Data/DataFrameColumn.BinaryOperations.cs [hunk 6]
```diff
@@ -12,96 +12,248 @@ namespace Microsoft.Data
 {
     public abstract partial class DataFrameColumn
     {
+        public static DataFrameColumn Add<T>(T value, DataFrameColumn right)
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrameColumn.BinaryOperations.cs]: I'm confused why we have both public static `Add(T, DataFrameColumn)` and `DataFrameColumn.RAdd(T)`? Why do they both need to be public?

File: src/Microsoft.Data/DataFrame.BinaryOperations.cs (hunk 4 above)
Comment (pgovind) [reply src/Microsoft.Data/DataFrame.BinaryOperations.cs]: Hmmm, I thought `RAdd` followed our guidelines. Are you looking at the "favor readability over brevity" one? I could live with `ReverseAdd` I guess. This is my comment from before:

> `RAdd` follows our naming guidelines, but `ROr` sounds a little weird. `ReverseAdd/ReverseOr` sound better?

Comment (eerhardt) [reply src/Microsoft.Data/DataFrame.BinaryOperations.cs]: ✓ DO favor readability over brevity.

> The property name CanScrollHorizontally is better than ScrollableX (an obscure reference to the X-axis).

What is `R` in these methods? Just like the "obscure reference to the X-axis" above, it is hard to guess what is meant by a single letter.

And yes:

X DO NOT use abbreviations or contractions as part of identifier names.

Also applies here.

Comment (eerhardt) [reply src/Microsoft.Data/DataFrame.BinaryOperations.cs]: I would spell out `Reverse` and keep it before the operation name.

`ReverseAdd`, `ReverseSubtract`, `ReverseDivide`, etc.

File: src/Microsoft.Data/DataFrameColumn.BinaryOperations.cs (hunk 6 above)
Comment (pgovind) [reply src/Microsoft.Data/DataFrameColumn.BinaryOperations.cs]: I decided to get rid of the static methods. The static operators now call the appropriate instance methods directly.

File: src/Microsoft.Data/PrimitiveColumnContainer.BinaryOperations.cs [hunk 7]
```diff
@@ -190,5 +190,45 @@ public PrimitiveColumnContainer<T> ElementwiseLessThan(T scalar, PrimitiveColumn
             return this;
        }
 
+       public PrimitiveColumnContainer<T> RAdd(T scalar)
```
Comment (pgovind) [inline src/Microsoft.Data/PrimitiveColumnContainer.BinaryOperations.cs]: This should change to Reverse

**Comment by pgovind** [comment]
Rebased on top of master

File: src/Microsoft.Data/DataFrame.BinaryOperators.cs [hunk 8]
```diff
@@ -16,339 +16,682 @@ public partial class DataFrame
         {
             return df.Add(value);
         }
+        
+        public static DataFrame operator +(bool value, DataFrame df)
+        {
+            return df.ReverseAdd(value);
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrame.BinaryOperators.cs]: I probably should have caught this months ago, but we have a chance of a null reference exception here if `df` is null. This also is possible in the existing code above. Maybe we can just log a separate bug for it and fix it later.

ex.

```
DataFrame df = null;
df2 = 1 + df;
```

I'm not entirely sure on what the behavior here should be. If we should throw an exception, or return `null`. Thoughts?

File: src/Microsoft.Data/PrimitiveDataFrameColumn.ReversedBinaryOperations.tt [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+                        // No conversions
+                        PrimitiveDataFrameColumn<U> primitiveColumn = this as PrimitiveDataFrameColumn<U>;
+                        PrimitiveDataFrameColumn<U> newColumn = inPlace ? primitiveColumn : primitiveColumn.Clone();
+                        newColumn._columnContainer.Reverse<#=method.MethodName#>(value);
+                        return newColumn;
+                    }
+                    else
+                    {
+                        PrimitiveDataFrameColumn<decimal> clonedDecimalColumn = CloneAsDecimalColumn();
+                       clonedDecimalColumn._columnContainer.Reverse<#=method.MethodName#>(DecimalConverter<U>.Instance.GetDecimal(value));
```
Comment (eerhardt) [inline src/Microsoft.Data/PrimitiveDataFrameColumn.ReversedBinaryOperations.tt]: (nit) white space doesn't line up here.

File: src/Microsoft.Data/PrimitiveDataFrameColumn.ReversedBinaryOperations.cs [hunk 10]
```diff
@@ -0,0 +1,394 @@
+﻿
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (eerhardt) [inline src/Microsoft.Data/PrimitiveDataFrameColumn.ReversedBinaryOperations.cs]: Oops - the rebase left these in the old folder.

Comment (pgovind) [reply src/Microsoft.Data/PrimitiveDataFrameColumn.ReversedBinaryOperations.cs]: Yup, I see it. Fixing it

File: src/Microsoft.Data/DataFrame.BinaryOperators.cs (hunk 8 above)
Comment (pgovind) [reply src/Microsoft.Data/DataFrame.BinaryOperators.cs]: We should throw an exception. Filed https://github.com/dotnet/corefxlab/issues/2776

**Pull request merged**